numpy/_core/src/multiarray/nditer_constr.c

```diff
@@ -1931,6 +1923,390 @@ operand_different_than_broadcast: {
     }
 }
 
+
+/*
+ * At this point we (presumably) use a buffered iterator and here we want
+ * to find out the best way to buffer the iterator in a fashion that we don't
+ * have to figure out a lot of things on every outer iteration.
+ *
+ * How do we iterate?
+ * ------------------
+ * There are currently two modes of "buffered" iteration:
+ * 1. The normal mode, where we either buffer each operand or not and
+ *    then do a 1-D loop on those buffers (or operands).
+ * 2. The "reduce" mode.  In reduce mode (ITFLAG_REDUCE) we internally use a
+ *    a double iteration where for "reduce" operands we have:
+ *      - One outer iteration with stride == 0 and a core with at least one
+ *        stride != 0 (all of them if this is a true reduce/writeable operand).
+ *      - One outer iteration with stride != 0 and a core of all strides == 0.
+ *    This setup allows filling the buffer with only the stride != 0 and then
+ *    doing the double loop.
+ *
+ * Only a writeable (reduce) operand require this reduce mode because for
+ * reading it is OK if the buffer holds duplicates elements.
+ * The goal of the reduce mode is that it allows for larger core sizes and
+ * buffers since the zero strides do not allow a single 1-d iteration.
+ * If do we use the reduce-mode, we can apply it also to read-only operands
+ * as an optimization.
+ *
+ * The functino here finds an outer dimension and it's "core" to use that
+ * works with reductions.
+ * While iterating, we will fill the buffers making sure that we:
+ *   - Never buffer beyond the outer dimension (optimize chance of re-use).
+ *   - Never buffer beyond the end of the core (all but outer dimension)
+ *     if the user manually moved the iterator to the middle of the core.
+ *     (Not typical and if, should usually happen once.)
```

Comment:
Restructured the order and wording to make it clearer hopefully.

You always have to (re-)fill buffers when manually moved (or ranged).

The point is that I decided to just finish a single core (which could be very small).  For reductions, you have to do that anyway.  Otherwise, it just seems a bit simpler to not worry about what happens if you iterator 1.5 core-sizes (rather than N core-sizes).